test/ApplicationTest.php

```diff
@@ -19,37 +22,56 @@ public function setUp()
     {
         $events = new EventManager();
 
-        $request = $this->getMockBuilder('Zend\Http\PhpEnvironment\Request')
-            ->disableOriginalConstructor()
-            ->getMock();
-        $response = $this->getMockBuilder('Zend\Http\PhpEnvironment\Response')
-            ->disableOriginalConstructor()
-            ->getMock();
+        $request = $this->prophesize(PhpEnvironment\Request::class);
+        $response = $this->prophesize(PhpEnvironment\Response::class);
 
-        $services = $this->getMockBuilder('Zend\ServiceManager\ServiceManager')
-            ->disableOriginalConstructor()
-            ->getMock();
+        $services = $this->prophesize(ServiceManager::class);
         $this->services = $services = $this->setUpServices($services, $events, $request, $response);
 
-        $app = new Application([], $services);
+        $app = $this->createApplication($services, $events, $request, $response);
         $this->app = $app = $this->setUpMvcEvent($app, $request, $response);
     }
 
-    public function setUpServices($services, $events, $request, $response)
+    /**
+     * Create and return an Application instance.
+     *
+     * Checks to see which version of zend-mvc is present, and uses that to
+     * determine how to construct the instance.
+     *
+     * @param ServiceManager $services
+     * @param EventManager $events
+     * @param PhpEnvironment\Request $request
+     * @param PhpEnvironment\Response $response
```

Comment:
So maybe both? `ObjectProphecy|ServiceManager` etc

It's not confusing, because you can't pass just `ServiceManager`.